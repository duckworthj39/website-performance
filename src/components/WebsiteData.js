import React, { useState, useEffect } from 'react';
import BarChart from './BarChart';
import LoadingSpinner from './LoadingSpinner';

const WebsiteData = () => {
  const [amazonData, setAmazonData] = useState({});
  const [bestBuyData, setBestBuyData] = useState({});
  const [targetData, setTargetData] = useState({});
  const [walmartData, setWalmartData] = useState({});
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    // Fetch data from the express endpoint
    // In a normal app you would configure the endpoint using env vars or a config file
    fetch(`http://localhost:3001/api/pagespeed`)
      .then((response) => response.json())
      .then((data) => {
        // Find each sites data and set it to state
        setAmazonData(
          data.find(function (item) {
            return item.site === 'https://www.amazon.com';
          })
        );
        setBestBuyData(
          data.find(function (item) {
            return item.site === 'https://www.bestbuy.com';
          })
        );
        setTargetData(
          data.find(function (item) {
            return item.site === 'https://www.target.com';
          })
        );
        setWalmartData(
          data.find(function (item) {
            return item.site === 'https://www.walmart.com';
          })
        );
        setIsLoading(false);
      })
      .catch((error) => console.log(error));
  }, []);

  if (isLoading) {
    return (
      <div>
        <LoadingSpinner />
      </div>
    );
  }

  return (
    <div>
      <BarChart
        labels={['Amazon', 'Walmart']}
        data={[amazonData.speedScore, walmartData.speedScore]}
        label="Performance Walmart"
        backgroundColor="rgba(255, 99, 132, 0.2)"
        borderColor="rgba(255, 99, 132, 1)"
        borderWidth={1}
      />
      <BarChart
        labels={['Amazon', 'BestBuy']}
        data={[amazonData.speedScore, bestBuyData.speedScore]}
        label="Performance Best Buy"
        backgroundColor="rgba(54, 162, 235, 0.2)"
        borderColor="rgba(54, 162, 235, 1)"
        borderWidth={1}
      />
      <BarChart
        labels={['Amazon', 'Target']}
        data={[amazonData.speedScore, targetData.speedScore]}
        label="Performance Target"
        backgroundColor="rgba(255, 206, 86, 0.2)"
        borderColor="rgba(255, 206, 86, 1)"
        borderWidth={1}
      />
    </div>
  );
};
export default WebsiteData;
