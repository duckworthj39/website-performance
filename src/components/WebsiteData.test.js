import React from 'react';
import { act, render } from '@testing-library/react';
import WebsiteData from './WebsiteData';
import mockBarChart from './BarChart';

jest.mock('./BarChart', () => jest.fn());

describe('WebsiteCharts', () => {
  const mockData = [
    { site: 'https://www.amazon.com', speedScore: 0.9 },
    { site: 'https://www.bestbuy.com', speedScore: 0.8 },
    { site: 'https://www.target.com', speedScore: 0.7 },
    { site: 'https://www.walmart.com', speedScore: 0.6 }
  ];

  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue(mockData)
    });
  });

  beforeEach(() => {
    jest.resetModules();
    jest.mock('./BarChart', () => mockBarChart);
  });

  afterEach(() => {
    jest.unmock('./BarChart');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render the loading spinner when the data is loading', () => {
    jest.spyOn(global, 'fetch').mockRejectedValueOnce(new Error('Something went wrong'));
    const { getByTestId } = render(<WebsiteData />);
    const loadingSpinner = getByTestId('loading-spinner');
    expect(loadingSpinner).toBeInTheDocument();
  });

  it('should render the BarChart components after the data is loaded', async () => {
    await act(async () => {
      render(<WebsiteData />);
    });
    expect(mockBarChart).toHaveBeenCalledTimes(3);
  });
});
