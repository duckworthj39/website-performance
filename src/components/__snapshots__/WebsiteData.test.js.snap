// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsiteCharts should match the snapshot 1`] = `ShallowWrapper {}`;
